Use new PCS

xldenis opened the pull request:
This is based on my other branch updating rustc, but it will integrate Zack's new PCS in the stead of the old one. After restoring compilation I will extend this to support a Prusti 1.0 level of borrowing.

Commit:
Remove old pcs

Commit:
Add new pcs as submodule

Commit:
Mostly fix impure stuff

Commit:
Integrate NeoPCS into current impure encoding

Commit:
Fix some crashes

Aurel300 commented on prusti-encoder/src/encoders/type/predicate.rs:
```diff
@@ -691,7 +691,10 @@ impl<'vir, 'tcx> PredicateEncValues<'vir, 'tcx> {
                 .field_snaps_to_snap
-                .apply(self.vcx, &[inner_snap, self_ref])
+                .apply(self.vcx, &[inner_snap, 
+                    // TODO: Re-enable this once its purpose is understood
+                    // self_ref
```
The idea was, as the comment explains, that the snapshot of a mutable reference consists of both a snapshot of the pointee at the time of creation (IIRC) and the actual reference `Ref`. The latter is somewhat analogous to a prophecy. @JonasAlaif would have more info.

xldenis replied:
Yea I conceptually figured that but since it wasn't typing I needed to do more digging

Commit:
Basic borrowing

Aurel300 committed:
wip

Aurel300 committed:
wip: update to new PCS types

Aurel300 committed:
type encoding cleanup wip

Aurel300 committed:
(throw away commit)

Aurel300 committed:
snapshot encoding with domain builder

Aurel300 committed:
wip: predicate encoding with builder

Aurel300 committed:
wip 1

Aurel300 committed:
update viper toolchain, add wands and packages

Aurel300 committed:
wip 2

Aurel300 committed:
handle Assert terminator messages and map error spans

Aurel300 committed:
fix generics issues, support box derefs

Aurel300 committed:
add types to lazy exprs

Aurel300 committed:
fixup box

Aurel300 committed:
encode copy operands partially with impure, partially with pure

JonasAlaif committed:
Basic support for loops

Aurel300 committed:
move to next2, start extracting wand encoder

Aurel300 committed:
use all PCG phases, fold on RemoveEdge

Aurel300 committed:
apply wands

Aurel300 committed:
small format fix

JonasAlaif committed:
Fixes

JonasAlaif committed:
Clean up some of the code and some other fixes

JonasAlaif committed:
Add packaging wands for loops

JonasAlaif committed:
Fix `make_concrete/generic` handling

JonasAlaif committed:
Apply loop wands and fix labels in package scripts

JonasAlaif committed:
Package wands correctly: according to pcg annotations

JonasAlaif committed:
Fix wands in signatures

JonasAlaif committed:
Improve fn signature wands

JonasAlaif committed:
Fix a few small issues

JonasAlaif committed:
Add handling of edge path conditions (conditional apply/fold)

JonasAlaif committed:
Simplify prusti builtin handling (was broken on cargo-prusti)

JonasAlaif committed:
Update pcs -> pcg

JonasAlaif committed:
Fix CI with submodules

JonasAlaif committed:
Fix some CI and small improvements

JonasAlaif committed:
Update rustc

JonasAlaif committed:
Fix well-definedness of specs for generic pure fns

The type specs need to go first, so that subsequent user specs can `mk_concrete` the arguments/result

JonasAlaif committed:
Fix incorrect usize ops and add Uint bounds + validity

Aurel300 committed:
wip cleanup

Aurel300 committed:
wip cleanup

JonasAlaif committed:
Fix issue of wrong types used in primitive ops

Aurel300 committed:
wip cleanup

Aurel300 committed:
fmt
